### PR TITLE
mds: some optimizations around client Capability and Lease tracking

### PR DESCRIPTION
**Pull request by gardran**
This is a partial revival of @ukernel's optimization PR: https://github.com/ceph/ceph/pull/43126
which has 
- reducing RAM consumption in Capability tracking.
- doing some performance optimization in ClientLease tracking. 
- optimize sending cap messages by using target session directly instead of looking it up through Inode's client_caps.

The last commit from myself switches ClientLease tracking from std::map to boost::intrusive::set to reduce amount of lookups over client_leases collection .

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/mds/SnapRealm.h [hunk 1]
```diff
@@ -129,7 +129,7 @@ struct SnapRealm {
   std::set<SnapRealm*> open_children;    // active children that are currently open
 
   elist<CInode*> inodes_with_caps;             // for efficient realm splits
-  std::map<client_t, xlist<Capability*>* > client_caps;   // to identify clients who need snap notifications
+  std::map<client_t, elist<Capability*> > client_caps;   // to identify clients who need snap notifications
```
Comment (vshankar) [inline src/mds/SnapRealm.h:132]: This is a nice saving. Just for information sake, here is the size difference:

> 16   elist<>::item
> 32   xlist<>::item

This is from a sample MDS log (main branch build) that dumps struct sizes on start. 👍

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66850](https://tracker.ceph.com/issues/66850).

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/67089](https://tracker.ceph.com/issues/67089).

**Comment by vshankar** [comment]
I'm seeing some new failures in the branch which this PR is a part of. Trying to isolate the problematic change. Will update when done.

**Comment by gardran** [comment]
jenkins test api

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-vshankar-testing-20240801065302-debug

**Pull request merged**